Improve PortableGenericRecordBuilder missing fields handling

Using `PortableGenericRecordBuilder` all the fields from the `ClassDefinition` must be written. In case they are not, an exception is thrown, however it does not help the user to spot the exact cause of the problem.

```
com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.internal.serialization.impl.portable.PortableGenericRecord'
	at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115) ~[hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:175) ~[hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:151) ~[hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:136) ~[hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:124)
        ...
Caused by: java.lang.NullPointerException
	at com.hazelcast.internal.serialization.impl.portable.PortableGenericRecord.readInt(PortableGenericRecord.java:101) ~[hazelcast-4.1-SNAPSHOT.jar:4.1-SNAPSHOT]
```

An exception with proper name and/or description should be used.

--- a/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableGenericRecordBuilder.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableGenericRecordBuilder.java
@@ -29,25 +29,39 @@ public class PortableGenericRecordBuilder implements GenericRecord.Builder {
 
     private final ClassDefinition classDefinition;
     private final Object[] objects;
-    private final boolean[] isOverWritten;
+    private final boolean[] isWritten;
     private final boolean isClone;
 
     public PortableGenericRecordBuilder(ClassDefinition classDefinition) {
         this.classDefinition = classDefinition;
         this.objects = new Object[classDefinition.getFieldCount()];
         this.isClone = false;
-        this.isOverWritten = null;
+        this.isWritten = new boolean[objects.length];
     }
 
     PortableGenericRecordBuilder(ClassDefinition classDefinition, Object[] objects) {
         this.classDefinition = classDefinition;
         this.objects = objects;
         this.isClone = true;
-        this.isOverWritten = new boolean[objects.length];
+        this.isWritten = new boolean[objects.length];
     }
 
+    /**
+     * @return newly created GenericRecord
+     * @throws HazelcastSerializationException if a field is not written when building with builder from
+     *                                         {@link GenericRecord.Builder#portable(ClassDefinition)} and
+     *                                         {@link GenericRecord#newBuilder()}
+     */
     @Override
     public GenericRecord build() {
+        if (!isClone) {
+            for (int i = 0; i < isWritten.length; i++) {
+                if (!isWritten[i]) {
+                    throw new HazelcastSerializationException("All fields must be written when building"
+                            + " a GenericRecord for portable, unwritten field :" + classDefinition.getField(i));
+                }
+            }
+        }
         return new PortableGenericRecord(classDefinition, objects);
     }
 
@@ -154,17 +168,15 @@ public class PortableGenericRecordBuilder implements GenericRecord.Builder {
     private GenericRecord.Builder write(String fieldName, Object value, FieldType fieldType) {
         FieldDefinition fd = check(fieldName, fieldType);
         int index = fd.getIndex();
-        if (objects[index] != null) {
+        if (isWritten[index]) {
             if (!isClone) {
                 throw new HazelcastSerializationException("It is illegal to the overwrite the field");
-            } else if (isOverWritten[index]) {
+            } else {
                 throw new HazelcastSerializationException("Field can only overwritten once with `cloneWithBuilder`");
             }
         }
         objects[index] = value;
-        if (isClone) {
-            isOverWritten[index] = true;
-        }
+        isWritten[index] = true;
         return this;
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecord.java
+++ b/hazelcast/src/main/java/com/hazelcast/nio/serialization/GenericRecord.java
@@ -275,6 +275,7 @@ public interface GenericRecord {
 
         /**
          * @return a new constructed GenericRecord
+         * @throws HazelcastSerializationException when the GenericRecord cannot be build.
          */
         GenericRecord build();
 

--- a/hazelcast/src/test/java/com/hazelcast/genericrecord/GenericRecordBuilderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/genericrecord/GenericRecordBuilderTest.java
@@ -101,4 +101,20 @@ public class GenericRecordBuilderTest {
             }
         });
     }
+
+    @Test
+    public void testUnwrittenFieldsThrowException() {
+        ClassDefinition classDefinition =
+                new ClassDefinitionBuilder(TestSerializationConstants.PORTABLE_FACTORY_ID, TestSerializationConstants.NAMED_PORTABLE)
+                        .addUTFField("name").addIntField("myint").build();
+
+        GenericRecord.Builder builder = GenericRecord.Builder.portable(classDefinition);
+        builder.writeInt("myint", 1);
+        assertThrows(HazelcastSerializationException.class, new Runnable() {
+            @Override
+            public void run() {
+                builder.build();
+            }
+        });
+    }
 }

--- a/hazelcast/src/test/java/com/hazelcast/genericrecord/GenericRecordBuilderTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/genericrecord/GenericRecordBuilderTest.java
@@ -44,12 +44,7 @@ public class GenericRecordBuilderTest {
         GenericRecord.Builder builder = GenericRecord.Builder.portable(namedPortableClassDefinition);
         builder.writeUTF("name", "foo");
         builder.writeInt("myint", 123);
-        assertThrows(HazelcastSerializationException.class, new Runnable() {
-            @Override
-            public void run() {
-                builder.writeUTF("name", "foo2");
-            }
-        });
+        assertThrows(HazelcastSerializationException.class, () -> builder.writeUTF("name", "foo2"));
     }
 
     @Test
@@ -64,12 +59,7 @@ public class GenericRecordBuilderTest {
                 .writeInt("myint", 123).build();
 
         GenericRecord.Builder builder = record.cloneWithBuilder().writeUTF("name", "foo2");
-        assertThrows(HazelcastSerializationException.class, new Runnable() {
-            @Override
-            public void run() {
-                builder.writeUTF("name", "foo3");
-            }
-        });
+        assertThrows(HazelcastSerializationException.class, () -> builder.writeUTF("name", "foo3"));
     }
 
     @Test
@@ -79,12 +69,7 @@ public class GenericRecordBuilderTest {
                         .addUTFField("name").addIntField("myint").build();
 
         GenericRecord.Builder builder = GenericRecord.Builder.portable(classDefinition);
-        assertThrows(HazelcastSerializationException.class, new Runnable() {
-            @Override
-            public void run() {
-                builder.writeUTF("nonExistingField", "foo3");
-            }
-        });
+        assertThrows(HazelcastSerializationException.class, () -> builder.writeUTF("nonExistingField", "foo3"));
     }
 
     @Test
@@ -94,12 +79,7 @@ public class GenericRecordBuilderTest {
                         .addUTFField("name").addIntField("myint").build();
 
         GenericRecord.Builder builder = GenericRecord.Builder.portable(classDefinition);
-        assertThrows(HazelcastSerializationException.class, new Runnable() {
-            @Override
-            public void run() {
-                builder.writeInt("name", 1);
-            }
-        });
+        assertThrows(HazelcastSerializationException.class, () -> builder.writeInt("name", 1));
     }
 
     @Test
@@ -110,11 +90,6 @@ public class GenericRecordBuilderTest {
 
         GenericRecord.Builder builder = GenericRecord.Builder.portable(classDefinition);
         builder.writeInt("myint", 1);
-        assertThrows(HazelcastSerializationException.class, new Runnable() {
-            @Override
-            public void run() {
-                builder.build();
-            }
-        });
+        assertThrows(HazelcastSerializationException.class, builder::build);
     }
 }

